src/CoreWCF.Kafka/src/CoreWCF/Channels/KafkaTransportPump.cs

```diff
@@ -0,0 +1,230 @@
+        }
+
+        foreach (var keyValue in _securityBindingElement.Config)
+        {
+            config.Set(keyValue.Key, keyValue.Value);
+        }
+
+
+        var (enableAutoCommit, enableAutoOffsetStore) = GetCommitStrategyConfigValues(config, _kafkaDeliverySemantics);
+        config.EnableAutoCommit = enableAutoCommit;
+        config.EnableAutoOffsetStore = enableAutoOffsetStore;
+
+        return config;
+    }
+
+    private static (bool? EnableAutoCommit, bool? EnableAutoOffsetStore) GetCommitStrategyConfigValues(ConsumerConfig consumerConfig, KafkaDeliverySemantics kafkaDeliverySemantics) =>
+        (kafkaDeliverySemantics, consumerConfig) switch
+        {
+            // KafkaBinding
+            (KafkaDeliverySemantics.AtMostOnce, { EnableAutoCommit: null, EnableAutoOffsetStore: null }) => s_atMostOnceConfigValues,
+            (KafkaDeliverySemantics.AtLeastOnce, { EnableAutoCommit: null, EnableAutoOffsetStore: null } ) => s_atLeastOnceBatchCommitConfigValues,
+            // CustomBinding
+            (KafkaDeliverySemantics.AtMostOnce, { EnableAutoCommit: false, EnableAutoOffsetStore: null }) => s_atMostOnceConfigValues,
+            (KafkaDeliverySemantics.AtLeastOnce, { EnableAutoCommit: true, EnableAutoOffsetStore: false } ) => s_atLeastOnceBatchCommitConfigValues,
+            (KafkaDeliverySemantics.AtLeastOnce, { EnableAutoCommit: false, EnableAutoOffsetStore: null }) => s_atLeastOncePerMessageCommitConfigValues,
+            _ => throw new NotSupportedException(string.Format(SR.ProvidedKafkaConfigurationIsNotValid, kafkaDeliverySemantics, consumerConfig.EnableAutoCommit, consumerConfig.EnableAutoOffsetStore))
+        };
+
+    public override Task StartPumpAsync(QueueTransportContext queueTransportContext, CancellationToken token)
+    {
+        _cts = new();
+        _mres = new();
+        _mres.Reset();
+
+        CancellationToken ct = CancellationTokenSource
+            .CreateLinkedTokenSource(_cts.Token, token).Token;
+
+        ConsumerConfig config = BuildConfig();
+        _consumer = new ConsumerBuilder<Null, byte[]>(config)
+            .SetKeyDeserializer(Deserializers.Null)
+            .SetValueDeserializer(Deserializers.ByteArray)
+            .SetLogHandler(OnLog)
+            .SetErrorHandler(OnError)
+            .Build();
+
+        _consumer.Subscribe(_topic);
+
+        Task.Run(async () =>
+        {
+            while (!ct.IsCancellationRequested)
+            {
+                try
+                {
+                    var consumeResult = _consumer.Consume(ct);
+                    if (config.EnablePartitionEof == true && consumeResult.IsPartitionEOF)
+                    {
+                        continue;
+                    }
+
+                    _logger.LogInformation("Received message from kafka at {topicPartitionOffset}", consumeResult.TopicPartitionOffset);
+                    if (_kafkaDeliverySemantics == KafkaDeliverySemantics.AtMostOnce)
+                    {
```

Comment:
@g7ed6e I also had a conversation with @mconnew to capture failure scenarios... we can probably expose a 1) DLQ in the kafkatransportbinding which if populated can be used to send failed message or 2) we can expose topic where failed message can be written... problem with (1) is we don't know what is the Queue used based on that logic would vary. Problem with (2) is we will bring Producer side dependency to the Corewcf service(consumer side)...  (3) option would be provide an extension method for user to handle via Kafka transport... let me know what you and @mconnew think.. if kafka provides out of the boox support for DLQ let me know, I couldn't find that.

Reply:
I don't like the idea of having an extension method, it's too high of a bar of how much code someone would need to write to use the feature. It should just be doable by configuring the binding.  
Having a very quick look over the Confluent Kafka documentation, it looks like there's no additional dependencies between consumer and producer. [They both use](https://developer.confluent.io/get-started/dotnet/#create-project) the Confluent.Kafka package.

Reply:
@mconnew I am fine with option (2)... let's see @g7ed6e thinks..

Reply:
@birojnayak / @mconnew afaik DLQ is not supported OOB by kafka. the easiest way indeed is to add a producer. For the time being i'm just discarding events to avoid blocking the consumer and trigger a rebalance.

Reply:
> shouldn't we commit once we process successfully ? (I do understand it's debatable.. if not we need to expose a way for user to act something suggested [here](https://www.confluent.io/en-gb/blog/spring-kafka-can-your-kafka-consumers-handle-a-poison-pill/#dead-letter-topic) if it fails. Else it would be hard for debugging.

@birojnayak the preferred Confluent.Kafka .NET way to implement at most once is described [here](https://docs.confluent.io/kafka-clients/dotnet/current/overview.html#consumer). Basically we need to
- disable autooffsetstore and manually call StoreOffsets
- enable autocommit